_posts/grant_guo/2018-03-22-kf-etl.md

```diff
@@ -0,0 +1,349 @@
+
+The `Try` class is defined for this purpose. A `Try[T]` represents a computation that may result in a value of type `T`. If the computation is successful, a `Success[T]` is returned. Or a `Failture[T]` is returned. We can make use of the pattern matching as the following:
+```
+val computation = Try(....)
+computation match {
+    case Success(ret) => { logic for successful operation }
+    case Failture(failed) => { error handling } 
+}
+
+```
+
+## Implementation Details
+
+This section explains how the ETL applies Scala's functional features.
+<figure>
+    <img src="{{site.urlimg}}grant_guo/pipeline.png" />
+    <figcaption>Kids-First ETL Pipeline</figcaption>
+</figure>
+
+### Data Models
+
+The ETL uses [ScalaPB](https://scalapb.github.io/), a Scala implementation of Google [Protocol Buffers](https://developers.google.com/protocol-buffers/), as the representation of the internal data model. [Protocol Buffers](https://developers.google.com/protocol-buffers/) are  language-neutral, platform-neutral extensible mechanisms for serializing structured data. It ensures that all exchanged data follows predefined formats and increases  data consistency.
+
+ScalaPB is built on top of the Protocol Buffer compiler and has many advantages that the ETL values including:
+* Supports both proto2 and proto3. The ETL uses `proto2` because some fields of the source data are optional and only `proto2` supports the keywords `optional` and `required`. 
+* Supports SparkSQL
+* Supports converting to and from JSON. The ETL uses this to directly generate JSON strings.
+* Supports generating case classes for all messages. Applying case classes to Spark Dataset is intuitive, so that no extra [Encoder](http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.Encoder)s are needed.
+* Supports introspection of generated case classes 
+
+### Context
+
+`Context` holds globally shared resources like `SparkSession`, `PostgresQL`, and `Elasticsearch` and is defined like this:
+
+```
+object Context extends ContextTrait with ClasspathURLEnabler{
+  lazy val (hdfs, rootPath) = getHDFS()
+  lazy val sparkSession = getSparkSession()
+  lazy val postgresql = getPostgresql()
+  lazy val esClient = getESClient()
+  
+  ...
+}
+```
+
+### Processor
+
+`Processor` is the key abstraction in the ETL and is defined like this:
+```
+trait Processor[I, O] extends Function1[I, O]{
+  override def apply(input: I): O = {
+    process(input)
+  }
+  def process(input:I): O
+}
+```
+
+The `Processor` is derived from Scala's [Function1[-T1, +R]](http://www.scala-lang.org/api/2.11.12/#scala.Function1) which basically means that a `Processor` is just a function. Any concrete `Processor` fills its logic by overriding the function `process(input:I):O`. Technically, a `Processor` could be implemented in any direct or indirect way. However, the reference implementations follow this convention:
```

Comment:
"Technically, a `Processor` could be implemented in any direct or indirect way."
I don't understand this sentence. :-(